### DISCUSSION TITLE
Requirements for scheduled tasks

### DISCUSSION THREAD

**Discussion by duckbytes** [platelet-app/platelet#149] [Ideas]
**Possible data to be captured for scheduled tasks**

- Time of pick up. before, on or after?
- Time of drop off. before, on or after?
- Date of pick up, if it's on a different day.

Fields to be added to the guided setup process, for coordinators to schedule live tasks during a shift. Editable fields to be added the task overview.

**Recurring scheduled tasks**

Should only be added or changed by admins.

Frequency:
- Daily, weekly, or monthly.
- Daily needs at least the time of day.
- Weekly needs at least the day of the week.
- Monthly needs at least the day of the month, but not the day of the week.

Weekly or monthly jobs don't require a time of day (could be done any time that day).

At the moment our only requirement is for daily air ambulance deliveries. So we could only implement daily tasks for now.

Likely a cron string will be the easiest way to store this information in the database.

A task template will need to be added for recurring scheduled tasks. I think this can be identical to the current process for adding a new task. The only exception might be that comments/notes aren't included, because there is no author. If this is required some workaround could be implemented.

The templated tasks could use the same model as current Tasks, but they will need to be filtered from the dashboard. Alternatively a stripped down model could be added to the database specifically for templated tasks. That record would be linked to a parent record that details the timing information.

Tasks will be created by a periodic lambda function. It will look up any records of active recurring scheduled tasks and add them to the system some hours in advance. Could be 12 hours in advance?

An "ends on" date could be added to stop recurring jobs at a particular date.

When a task is added by the lambda function, it will have no assignees. The system can't be aware of who is going to be responsible on the day, so the coordinator on shift will have to add themselves as a coordinator. This would involve selecting ALL from the role view selector, selecting all items and assigning to themselves.

In future this could be made easier by filters instead, especially if the role view selector workflow is removed. https://github.com/platelet-app/platelet/discussions/146

**Dashboard**

For scheduled tasks, the dashboard cards will need to indicate that it is scheduled at a specific time.  This is only necessary if there is a specific time to pick up or deliver, not if the task is any time that day.

This could be an icon or label. There might be too much possible pieces of information to place specific times on the card itself. The indicator on the card should tell the user to open the task overview to see specific timing information.

**Comment by Merseybiker** [comment]
Merseyside and Cheshire Blood Bikes have this requirement.  We have between 5 and 10 recurring jobs - some examples are below:

1) Each day district nurses drop off blood samples at an NHS surgery.  We pick them all up (sometimes 4 bags containing over 25 samples in each bag) and take them from the surgery to an NHS Trust Pathology Lab - this happens at 5pm every working day (Monday to Friday).

2) We run litres of milk from one Children's NHS Trust to another one every day at 12.30 (lunchtime) - typically between 4 and 6 litres of DBM (Donor Breast Milk).

These jobs are standard.  We also have recurring jobs sometimes when we have blood transfusion runs - so take a special container from a Marie Curie Hospice into Liverpool City (one job in Platelet), then pick up the bloods and run these back to Marie Curie (a 2nd job in Platelet).  These can be every day at the same time for a couple of weeks...

The jobs are allocated to a rider, but we need to create the template then be able to generate the jobs off the template - possibly having a frequency, pickup location, drop off location, start date, end date, and agreed pick up time (optional) would be good for these templates.  In addition things like blood transfusion runs have critical timings - so an optional SLA timer that shows the time we have available (countdown from pickup) would be very useful.

**Comment by duckbytes** [comment]
Some points to consider when implementing this feature:

- The coordinator who creates the task might not be the coordinator on the day
- Some tasks are recurring and some are not
- The guided setup will need to be updated to allow creating a future task
- Future tasks should not be presented to the current coordinator until the day it is due (or the day before?)
- When the task is due it should appear in the "pending" tab

The guided setup UI doesn't need to account for recurring tasks as that will be kept elsewhere. For now it can just show a date and time.

When saving the task the coordinator could be prompted to assign the task to themselves and place in NEW, or add the the queue instead.

When looking at the pending tab on the dashboard, a secondary set of tabs could be displayed to display either pending (due today) or "upcoming" tasks.

The due date, time, and other job details should be editable by coordinators from the upcoming tab.

**Comment by duckbytes** [comment]
Draft idea for a design. Add buttons on the guided setup to allow setting the "Schedule".

![Frame 4](https://github.com/user-attachments/assets/47e8fea7-2042-4c15-be5c-3054d997fba1)

This opens a dialog that lets the user specify "anytime, before, at, after" a specific time. Also allows the user to set the date. By default it is "anytime today".

![Frame 1(1)](https://github.com/user-attachments/assets/f80becfe-c72e-448c-9236-29c350520ee9)

Alternatively there could be one button to open one dialog that covers both pick up and drop off schedules.

Similar buttons to open the same dialog could be added to the task overview.

**Comment by Merseybiker** [comment]
Looks fine.  We'll obviously need to test this.  Main points
1) Controller on the day of the job needs to be able to see the new job.
2) It needs to away from the main dashboard till the day of the job.

**Reply by duckbytes** [reply]
I think the easiest way will be to add an "upcoming" column in the pending tab which will show future jobs. When a job is due within the next day, it will automatically be moved to the "pending" column. At this point a notification badge showing how many pending jobs there are will update so the user knows they are available.

The recurring jobs feature already does this partially and it works well.

**Reply by duckbytes** [reply]
One thing that maybe needs to be made clear is how the dashboard will react when the user adds a future job.

It could:

- Add the job to the future column if it's more than a day away and not assign the coordinator. The user will need some kind of notification so it doesn't seem to "disappear".
- Add the job to the future column and assign that coordinator automatically.
- On saving the job, prompt the user to choose between assigning themself and adding it to the NEW column or not assigning themself and adding it to the FUTURE column.

**Comment by Merseybiker** [comment]
We have about 20 or more controllers, and often may not know which controller to assign the job to until the day.  My view is to keep the job unallocated to a controller and allow all controllers to see the "Pending" jobs area.  The controller can then allocate themselves to the job on the day.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
